### PR TITLE
Add static mesh, new shapes, dynamic color and variable render resolution

### PR DESCRIPTION
**Pull request by phohenberger**
- Add option for meshes to be static, meaning they will only be loaded in once and not changed during the following frames. This significantly improves rendering time for a larger non-moving mesh. 
Can be added to `vis.Particle` and `vis.VectorField` by simply giving them the `static = True` parameter.
Note that static objects only take in position in the shape of `(n_part, n_dim)`
- Add 7 more mesh shapes. `Box`, `Cone`, `Icosahedron`, `Octahedron`, `Tetrahedron`, `Torus` and `MobiusLoop`
- Allow to choose the render resolution and samples per pixel for screenshots or exported videos. Default is `resolution = [4096, 2160]` and `spp = 64`

### CODE REVIEW

File: znvis/mesh/box.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+"""
+
+from dataclasses import dataclass
+
+import numpy as np
+import open3d as o3d
+
+from znvis.transformations.rotation_matrices import rotation_matrix
+
+from .mesh import Mesh
```
Comment (SamTov) [inline znvis/mesh/box.py:34]: Can you add the full import path for all of them mesh objects?

File: znvis/mesh/cone.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+"""
+
+from dataclasses import dataclass
+
+import numpy as np
+import open3d as o3d
+
+from znvis.transformations.rotation_matrices import rotation_matrix
+
+from .mesh import Mesh
```
Comment (SamTov) [inline znvis/mesh/cone.py:34]: Same comment

File: znvis/mesh/icosahedron.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
+"""
```
Comment (SamTov) [inline znvis/mesh/icosahedron.py:1]: After going over the third one, I am realizing that we could put these into a more comprehensive parent class. I think the `create_mesh` function could be generalized and we could use class propoerties e.g. for the kind of mesh, and dict-like attributes, or an automatically generated data-class (https://docs.python.org/3/library/dataclasses.html#dataclasses.make_dataclass) based on the parameters of the specific mesh. Alternatively, we make a generic method called `instantiate_mesh` which does the vertex computation, the translation, and the rotation, and we avoid re-writing 5 times or so. What do you think?

**Review by SamTov — CHANGES_REQUESTED** [review]
Great PR! I have a few comments on the mesh objects I would like to discuss briefly. But overall it looks great. 

On a second point, I think we can clean up the data flow in the visualiser and the renderer, however, I think this calls for a different PR so we can shelve that for now. Just check out my mesh comments and we can merge the PR.

**Comment by phohenberger** [comment]
I moved the mesh instantiation to the parent `mesh`-class. The child-classes now only need a `create_mesh` function which relies on the specific shape parameters. This reduces the repeated code significantly. What do you think @SamTov

**Comment by phohenberger** [comment]
Shapes like ´Arrow´ or ´CustomMesh´ are special cases and need to overwrite a function in the `mesh` parent-class. But i think this is still a good solution as they need to instantiate differently and that cannot be generalized.

**Comment by phohenberger** [comment]
The changes passed the tests locally.
```
================================== test session starts ==================================
platform linux -- Python 3.10.14, pytest-8.2.2, pluggy-1.5.0
rootdir: -
plugins: anyio-4.3.0, typeguard-4.2.1, dash-2.16.1
collected 13 items                                                                      

CI/integration_tests/test_simple_spheres.py .                                     [  7%]
CI/unit_tests/bounding_box/test_bounding_box.py ..                                [ 23%]
CI/unit_tests/mesh/test_cylinder.py ..                                            [ 38%]
CI/unit_tests/mesh/test_mesh.py .                                                 [ 46%]
CI/unit_tests/mesh/test_sphere.py ..                                              [ 61%]
CI/unit_tests/particle/test_particle.py ..                                        [ 76%]
CI/unit_tests/transformations/test_rotation_matrices.py .                         [ 84%]
CI/unit_tests/visualizer/test_visualizer.py ..                                    [100%]

================================== 13 passed in 14.73s ==================================
```

File: examples/all_shapes.py
Comment (SamTov) [inline examples/all_shapes.py:1]: This file needs a newline at the end. A linting run should handle that. There is pre-commit in this project, you can install it like:

```sh
precommit install .
```

File: znvis/mesh/mesh.py [hunk 4]
```diff
@@ -42,13 +42,15 @@ class Mesh:
...
     base_direction: np.ndarray = field(default_factory=lambda: np.array([1, 0, 0]))
+    dynamic_color: bool = False
 
     def __post_init__(self):
         """
         Post init function to create materials.
         """
         material = rendering.MaterialRecord()
-        material.base_color = np.hstack((self.material.colour, self.material.alpha))
+        if not self.dynamic_color:
```
Comment (SamTov) [inline znvis/mesh/mesh.py:52]: Is the bool necessary? The color will either be a tuple or a nested array right? Can we not just check the dimension of the array and determine if it is time-dependent?

**Review by SamTov — CHANGES_REQUESTED** [review]
I like the changes, I just have two comments which could be addressed quickly. And I like the new parent class updates.

**Comment by phohenberger** [comment]
Alright, i think now all the changes are adressed. @SamTov

**Review by SamTov — APPROVED** [review]
Thanks for the changes!

**Pull request merged**